### PR TITLE
Enable CORS requests on all routes

### PR DESCRIPTION
**Pull request by StrangeBytesDev**
Enable's CORS requests on all endpoints for all request types. Update's tests to reflect that change.

### CODE REVIEW

**Review by phymbert — APPROVED** [review]

**Review by ggerganov — CHANGES_REQUESTED** [review]
Lets put this feature behind a command-line flag: `--enable-cors`

By default, it should not be enabled

**Comment by Azeirah** [comment]
Ohhh so _that's_ how you apply middleware to routes, I couldn't find it at first.

I read the discussion in the linked thread, and I'm honestly a bit unsure about what the right approach is. In my opinion CORS is not an amazing security mechanism because as was already stated, anyone can still approach the server using anything other than a browser that respects CORS.

Still, you'd be preventing sites doing something similar to crypto mining on your pc by accessing any locally running LLMs if it's not disabled.

I want to add two points

The first is, if you do the flag thing, at the very least add a short log to the terminal that CORS is disabled when an endpoint is rejected due to CORS, informing the user they can enable it, but should understand the security risks (and recommend setting an API key)

So when CORS is disabled and an OPTIONS endpoint is hit, log a message to the console

> "Warning: site **example.com** is denied access the llama server API. Pass the --enable-cors flag to allow **example.com** to access this API. For security reasons, you are recommended to set an API key as well with the --api-key flag." 

---

The alternative is to enable it by default, but work on a whitelist basis. No hosts except localhost are allowed by default. You can enable hosts in something like a cors-whitelist.txt. This is how CORS is supposed to be used by design.

This is simpler and safer but requires people to interact with the whitelist file. It's possible to maybe add a Y/N prompt to the server when an unknown host is trying to access the server, or otherwise list a similar warning message as the one above

> "Warning: site **example.com** is denied access the llama server API. Add **example.com** to cors-whitelist.txt if you want it to have access to your llama.cpp API."

**Comment by StrangeBytesDev** [comment]
@Azeirah I love the log idea. I think that would substantially help cut down on confused users and new github issues if we disable CORS by default.

@ggerganov @ngxson thoughts on using a cors-whitelist .txt file, or adding a --cors-whitelist flag?

**Comment by ngxson** [comment]
@ggerganov @Azeirah Instead of disabling CORS by default, I think we should only allow a limited number of whitelist domain instead.

Currently, we are doing **both** CORS + checking `Origin` header  and it is **redundant**. What I can suggest is:

| Option | Origin | CORS | When user access to any endpoint (for example /completions) | Log |
| --- | --- | --- | --- | --- |
| by default | localhost | * | OK (access as normal) |  |
| by default | example.com | * | Error code 403: `{ "error": "Origin not whitelisted" }` | ERR: Origin example.com is not accepted |
| `--whilelist-domain example.com` | localhost | * | Error code 403: `{ "error": "Origin not whitelisted" }` | ERR: Origin localhost is not accepted |
| `--whilelist-domain example.com` | example.com | * | OK (access as normal) |  |
| `--whilelist-domain *` | (any) | * | OK (access as normal) | WARN: Allowing all origin without setting an API key, this is not recommended |

By throwing an error instead of letting browser to block it, some frontends can show the error message `Origin not whitelisted` in the UI.

**Comment by ngxson** [comment]
Also what we're currently doing:

```
res.set_header("Access-Control-Allow-Origin", req.get_header_value("Origin"));
```

This is equivalent to

```
res.set_header("Access-Control-Allow-Origin", "*");
```

That's why I say that it's redundant. What is interesting for us is to checkthat `Origin`

**Comment by Azeirah** [comment]
> Also what we're currently doing:
> 
> ```
> res.set_header("Access-Control-Allow-Origin", req.get_header_value("Origin"));
> ```
> 
> This is equivalent to
> 
> ```
> res.set_header("Access-Control-Allow-Origin", "*");
> ```
> 
> That's why I say that it's redundant. What is interesting for us is to checkthat `req.get_header_value("Origin")`

No, they're not equivalent. CORS requires the domain to be mirrored back explicitly in the case of credentialed requests (a credentialed request is one with the Authorization header)

**Comment by ngxson** [comment]
@Azeirah Ah yeah sorry I missed that part.

But even with what you say, then the fact that we currently mirror the `Origin` domain is just to allow credentialed requests for that specific domain right?

I this case, I think we should:
- Either set `Access-Control-Allow-Origin` to a pre-defined value, not reflecting `Origin` (the notion of whitelist as you suggested)
- Or, still allow `Access-Control-Allow-Origin` to reflect the Origin, but we internally check if `Origin` is in whitelist or not

Basically, what I suggest here is either use `Access-Control-Allow-Origin` as it supposed to be used (i.e. with a pre-defined whitelist), either checking if `Origin` is in whitelist, but **not both**.

**Comment by Azeirah** [comment]
> @Azeirah Ah yeah sorry I missed that part.
> 
> But even with what you say, then the fact that we currently mirror the `Origin` domain is just to allow credentialed requests for that specific domain right?
> 
> In this case, I think we should:
> 
>     * Either set `Access-Control-Allow-Origin` to a pre-defined value, not reflecting `Origin` (the notion of whitelist as you suggested)
> 
>     * Or, still allow `Access-Control-Allow-Origin` to reflect the Origin, but we internally check if `Origin` is in whitelist or not
> 
> 
> Basically, what I suggest here is either use `Access-Control-Allow-Origin` as it supposed to be used (i.e. with a pre-defined whitelist), either checking if `Origin` is in whitelist, but **not both**.

Yep! I think whitelisting is the best approach. The current implementation is a catchall

**Comment by ngxson** [comment]
> Yep! I think whitelisting is the best approach. The current implementation is a catchall

Thanks, we finally have an agreement here.

I modified my proposal, my idea is to keep the `Access-Control-Allow-Origin` reflect `Origin` (keep as-is), but check `Origin` header:

| Option | Origin | Access-Control-Allow-Origin | When user access to any endpoint (for example /completions) | Log |
| --- | --- | --- | --- | --- |
| by default | localhost | (reflect Origin) | OK (access as normal) |  |
| by default | example.com | (reflect Origin) | Error code 403: `{ "error": "Origin not whitelisted" }` | ERR: Origin example.com is not accepted |
| `--whilelist-domain example.com` | localhost | (reflect Origin) | Error code 403: `{ "error": "Origin not whitelisted" }` | ERR: Origin localhost is not accepted |
| `--whilelist-domain example.com` | example.com | (reflect Origin) | OK (access as normal) |  |
| `--whilelist-domain *` | (any) | (reflect Origin) | OK (access as normal) | WARN: Allowing all origin without setting an API key, this is not recommended |

By throwing an error instead of letting browser to block it, some frontends can show the error message `Origin not whitelisted` in the UI.

Do you still prefer this approach (reflect Origin) or you prefer to have `Access-Control-Allow-Origin` fixed to a pre-defined value? (i.e. via `--whilelist-domain`)

**Comment by Azeirah** [comment]
Only remark I have is to allow localhost regardless. I think it's a sensible default.

Am not 100% certain whether cors even applies to localhost though? So maybe not necessary. But many chat clients are just electron wrappers around some js webapp and they run into CORS issues.

**Comment by StrangeBytesDev** [comment]
I've implemented the feature as specified. The origin is always reflected. I added a flag "--public-domain" which allows users to specify a domain which will allow CORS requests. Any requests which specify an origin other than the public domain will receive an error message and an http 403 status. Blocked requests are logged. The public-domain setting defaults to "http://localhost:8080". Setting the public-domain to "*" without also including an API key logs a warning.

I used "public-domain" instead of "whitelist" because a CORS header only allows for setting a single origin (as far as I understand) and I implemented the origin check as a simple string comparison, so you can only specify a single CORS enabled origin. That seemed more robust, and if users want to enable multiple origins, then setting the public domain to "*" and using an API key is probably the best way to handle that.

File: examples/server/tests/features/security.feature [hunk 1]
```diff
@@ -47,5 +47,5 @@ Feature: Security
       | localhost       | Access-Control-Allow-Origin      | localhost         |
       | web.mydomain.fr | Access-Control-Allow-Origin      | web.mydomain.fr   |
       | origin          | Access-Control-Allow-Credentials | true              |
-      | web.mydomain.fr | Access-Control-Allow-Methods     | POST              |
+      | web.mydomain.fr | Access-Control-Allow-Methods     | *                 |
```
Comment (phymbert) [inline examples/server/tests/features/security.feature:50]: Actually can we just restrict to GET and POST ?

File: examples/server/README.md [hunk 2]
```diff
@@ -42,6 +42,7 @@ see https://github.com/ggerganov/llama.cpp/issues/1437
 - `-to N`, `--timeout N`: Server read/write timeout in seconds. Default `600`.
 - `--host`: Set the hostname or ip address to listen. Default `127.0.0.1`.
 - `--port`: Set the port to listen. Default: `8080`.
+- `--public-domain`: Set a public domain which will be allowed for Cross Origin Requests. If you are using the server as an API from a browser, this is required.
```
Comment (Azeirah) [inline examples/server/README.md:45]: Please rename it to reference CORS at the very least, public domain might confuse people something like `--cors-allowed-origin`.

```
- `--cors-origin`: Set what origin (example.com) is allowed to access the API. Use * to allow all origins (insecure).  If you are using the server as an API from a browser, this parameter is required.
```

File: examples/server/server.cpp [hunk 3]
```diff
@@ -2777,14 +2788,38 @@ int main(int argc, char **argv)
...
+
+        // Allow options so that request will return a specific error message rather than a generic CORS error
+        if (req.method == "OPTIONS") {
+            res.status = 200;
+            return httplib::Server::HandlerResponse::Handled;
+        }
+
+        if (req.has_header("Origin") && sparams.public_domain != "*") {
+            if (req.get_header_value("Origin") != sparams.public_domain) {
+                LOG_WARNING("Request from origin not allowed.", {{"origin", req.get_header_value("Origin")}});
```
Comment (Azeirah) [inline examples/server/server.cpp:2807]: Just wanted to note, I believe this also solves the issue that the server previously had, when a request was sent that was disallowed by CORS, the server would still process the request fully. So even disallowed origins could make your PC waste cycles by calling the completion API.

File: examples/server/README.md (hunk 2 above)
Comment (phymbert) [reply examples/server/README.md:45]: ++ and the user is sending cross origin requests.

Comment (phymbert) [reply examples/server/README.md:45]: also, to be coherent with other parameters,  I suggest `--http-cors-origin`

File: examples/server/README.md [hunk 4]
```diff
@@ -42,6 +42,7 @@ see https://github.com/ggerganov/llama.cpp/issues/1437
 - `-to N`, `--timeout N`: Server read/write timeout in seconds. Default `600`.
 - `--host`: Set the hostname or ip address to listen. Default `127.0.0.1`.
```
[the remark this answers is not in the document]
Comment (briancw) [reply examples/server/README.md:43]: There's currently a mixture of "127.0.0.1" and "localhost" throughout the codebase. I started to do some normalization to a single value, and moved towards localhost, as that's been the domain suggested by Azeirah and Ngxson. But then I ran into trouble with the python tests not liking localhost, so I went back, but missed this instance.
There's probably a consideration to be made about which value is more likely to be intuitive for a typical user, but maybe that's it's own issue.

[the remark this answers is not in the document]
Comment (StrangeBytesDev) [reply examples/server/README.md:43]: There's currently a mixture of "127.0.0.1" and "localhost" throughout the codebase. I started to do some normalization to a single value, and moved towards localhost, as that's been the domain suggested by Azeirah and Ngxson. But then I ran into trouble with the python tests not liking localhost, so I went back, but missed this instance.
There's probably a consideration to be made about which value is more likely to be intuitive for a typical user, but maybe that's it's own issue.

File: examples/server/tests/features/security.feature (hunk 1 above)
Comment (StrangeBytesDev) [reply examples/server/tests/features/security.feature:50]: I assume we want GET, POST, OPTIONS so that preflight requests are allowed.

File: examples/server/README.md (hunk 2 above)
Comment (StrangeBytesDev) [reply examples/server/README.md:45]: My thought was that a typical user probably wouldn't know what CORS is or why they need it. I've updated the flag to be "--http-cors-origin". There's potentially some consideration to avoiding "http" as the origin "https://example.org" is different from "http://example.org".

File: examples/server/server.cpp [hunk 5]
```diff
@@ -2132,6 +2134,15 @@ static void server_params_parse(int argc, char **argv, server_params &sparams,
...
             sparams.hostname = argv[i];
         }
+        else if (arg == "--http-cors-origin")
+        {
+            if (++i >= argc)
+            {
+                invalid_param = true;
+                break;
+            }
+            sparams.http_cors_origin = argv[i];
```
Comment (ngxson) [inline examples/server/server.cpp:2144]: Is it possible here to input a list of domain names, for example `example.com,mywebsite.com,example.net` ?

File: examples/server/server.cpp (hunk 3 above)
Comment (ngxson) [reply examples/server/server.cpp:2807]: The [Origin header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) actually have the protocol + port within it. I found a simple code snippet allow you to parse URI and extract the domain name here: https://gist.github.com/RedCarrottt/c7a056695e6951415a0368a87ad1e493

File: examples/server/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngxson) [reply examples/server/README.md:43]: I also prefer `127.0.0.1` because sometimes `localhost` is not defined in docker environment

File: examples/server/server.cpp (hunk 5 above)
Comment (StrangeBytesDev) [reply examples/server/server.cpp]: Done. It also made for a pretty convenient way to include both "localhost" and "127.0.0.1" by default.
